### DISCUSSION TITLE
proposal: Release Annotations

### DISCUSSION THREAD

**Discussion by lann** [bytecodealliance/SIG-Registries#11] [Ideas]
To support use cases N (publisher yanking package) and O (registry operator deleting package), I propose we define a new "Release Annotation" type which would have a many-to-one relationship with releases.

Example:
```jsonc
{
  "annotationType": "yank",
  "releaseId": "bytecodealliance.org:dog-facts/v1.0.0",
  // Creator of this annotation; permitted user roles depend on annotationType
  "creator": "dog-facts-publisher",
  // Creation time
  "createdAt": "2022-05-25T14:56:01Z",
  // Signature of this annotation; signed by "creator" key
  "signature": "...",
  // Optional (?) context for this annotation; a registry could require this to be a URI
  "reason": "https://github.com/bytecodealliance/dog-facts/issues/123",
}
```

Two "annotation types" could cover current use cases:

* `yank` (use case N)
  * An assertion by the publisher that "this release is not fit for use"
  * Creator must be an authorized publisher
  * Does not alter the release, but may cause the release to be excluded from default query results

* `takedown` (use case O)
  * An assertion by the publisher _or_ registry operator that "this release was removed for legal or policy reasons"
  * Creator may be an authorized publisher _or_ registry operator
  * Content URLs removed. Release metadata may be removed if the metadata itself violates policy

Other use-cases:

* Additional signatures (e.g. 3rd-party validation)

**Comment by michelleN** [comment]
@lann Does it make sense to include a timestamp in the release annotation type?

**Reply by lann** [reply]
Yes! Added.

**Reply by michelleN** [reply]
Given the two current annotation types, I may want to understand which one is the latest or most current to understand the current status of the release. Does it make sense to also add a key called latest and set that to true for which ever is the most recent annotation?

Timestamps are helpful but I'm wondering if they are enough and if there are cases where evaluating timestamps might not be enough?

The latest key/value may not make sense in the future if we want to add other annotation types but I'm curious on your thoughts around this.

**Reply by guybedford** [reply]
Yank and takedown seem like they would be irreversible so that future annotations would likely be redundant, apart from the case of initiating a takedown after a yank.

In the event that a publisher wants to send a message by running a takedown on a highly useful library of theirs that will break lots of things, it could be worth ensuring that all takedown requests do require registry review approval that they are in fact valid security issues, especially with high volume versions? Although that does mean having a completely responsive security role....

**Reply by lann** [reply]
I view `yank` as probably-sufficient for most security issues. You would perhaps use `takedown` for intentionally-malicious releases, but the primary use-case would be for _legal_ (DMCA et al) takedowns, where a judge might not accept the "but cryptography!" argument. Put differently, if a registry operator decides to delete a release I think we can't reasonably prevent that, but we can provide a more transparent alternative that makes the action obvious.

**Comment by guybedford** [comment]
Should we include the content hash in the release annotation, so that it is cryptographically associated with the exact release content without further API calls being necessary?

**Reply by lann** [reply]
Perhaps. If we go down the "immutable log" route it won't be necessary since it will "automatically" be cryptographically bound to the release entry.

**Reply by guybedford** [reply]
There is still value in being able to construct short proofs, even against an immutable log, without having to go through a large transaction history. For example, there might be a validation mode for APIs or a dedicated validation API that can return specific proofs about a package or version. Thinking about specific audit trails might still be useful.

**Comment by ricochet** [comment]
Will registry operators want to create policies when annotations like these are applied? E.g. takedowns require a link to where the DMCA's are escrowed.

```
reason(?string): "accidental-release"
description(?string): "See https://github.com/bytecodealliance/dog-facts/issues/123"
```

yankReason:
- accidental release
- severe bug
- unintentional breaking changes

takedownReason:
- DMCA

**Reply by lann** [reply]
Possibly! I can imagine some kind of "severity" associated with a yank could help clients make decisions about e.g. notifying users.

**Reply by ricochet** [reply]
Suggestion from @lann during today's meeting:

> "takedown" could also be a parameterization of "yank"
